src/Widgets/Modal/DownloadScenario.hs

```diff
@@ -55,36 +65,40 @@ downloadScenarioContent scB scContentPopupE = do
     elClass "div" "modal-footer" $
       elClass "p" "text-right" $ do
         ce <- buttonFlat "Close" def
-        se <- buttonFlat "Download" $ "onclick" =: "document['q$downloadScenarioContentUrl']();"
+        se <- buttonFlat "Download" $ "onclick" =: downloadScenario
         return $ leftmost [ce, se]
-  where
-    (runCode, nameFieldId) = $(do
-          nameFId <- newVar
-          scenarioContentUrl <- newVar
-          exportVars <- returnVars [nameFId]
-          let scurl = rawJS scenarioContentUrl
-          runCode' <- qjs
-            [julius|
-                var #{scurl} = null;
-                document['q$makeScenarioContentUrl'] = function(txt) {
-                    if (#{scurl} !== null) {
-                      window.URL.revokeObjectURL(#{scurl});
-                    }
-                    #{scurl} = window.URL.createObjectURL
-                        (new Blob([txt], {type: 'application/json;charset=utf-8'}));
-                };
-                document['q$downloadScenarioContentUrl'] = function() {
-                    var a = document.createElement("a");
-                    a.href = #{scurl};
-                    a.download = document.getElementById('#{rawJS nameFId}')['value'] + ".scenario";
-                    document.body.appendChild(a);
-                    a.click();
-                    document.body.removeChild(a);
-                };
-            |]
-          return $ ConE (tupleDataName 2) `AppE` runCode' `AppE` exportVars
-       )
 
+prepareFns :: MonadIO m => m Text
```

Comment:
Not good that we have to run IO action and re-define these two functions every time we need to get `nameFieldId`.
Maybe  we could put it into top-level `unsafePerformIO`?.. as for nameFId, we can pass it to `q$downloadScenarioContentUrl` as an argument.

Reply:
Not sure exactly what you're proposing... I guess the JSON serialization has to run before we can click the link? Or maybe we could do the serialization right in the `downloadScenarioContentUrl` function..?

Reply:
No, prepareFns does not do serialization, it only creates js functions for later use. That should be done only once, before they are needed for the first time.

Reply:
yes, we can move `prepareFns` somwhere else. Where exactly do you mean with "top-level `unsafePerformIO`"? (I could only find some `unsafePerformIO` in `Generations.hs`)

We'll have to run `prepareDownloadScenarioContent` every single time though...